File: p2p_android/app/src/main/java/com/example/p2p/data/remote/model/Dispute.kt
package com.example.p2p.data.remote.model

// ── Modelos de respuesta del backend ─────────────────────────────────────────

data class DisputeTransaction(
    val id: String,
    val amount_from: Double,
    val amount_to: Double,
    val exchange_rate: Double,
    val status: String,
    val buyer_id: String,
    val vendor_id: String
)

/**
 * Mapea exactamente la respuesta de:
 *   GET  /api/v1/disputes/my-disputes
 *   GET  /api/v1/disputes/{id}
 *   GET  /api/v1/admin/disputes
 *   GET  /api/v1/admin/disputes/{id}
 */
data class Dispute(
    val id: String,
    val transaction_id: String,
    val initiator_id: String,
    val initiator_name: String?,
    val reason: String,
    val description: String?,
    val status: String,                // open | under_review | resolved | closed
    val resolved_by: String?,
    val resolution: String?,           // favour_buyer | favour_vendor
    val resolution_note: String?,
    val resolved_at: String?,
    val created_at: String,
    val updated_at: String,
    val transaction: DisputeTransaction? = null
)

data class DisputesResponse(
    val disputes: List<Dispute>,
    val pagination: PaginationMeta? = null
)

data class PaginationMeta(
    val page: Int,
    val per_page: Int,
    val total: Int,
    val pages: Int,
    val has_next: Boolean,
    val has_prev: Boolean
)

// ── Requests ──────────────────────────────────────────────────────────────────

/**
 * Body para POST /api/v1/transactions/{id}/dispute
 */
data class CreateDisputeRequest(
    val reason: String,
    val description: String?
)

/**
 * Body para PATCH /api/v1/admin/disputes/{id}/resolve
 */
data class ResolveDisputeRequest(
    val resolution: String,           // "favour_buyer" | "favour_vendor"
    val resolution_note: String?
)

// ── Constantes de razón (deben coincidir con Dispute.VALID_REASONS en Python) ─

object DisputeReason {
    const val PAYMENT_NOT_RECEIVED = "payment_not_received"
    const val WRONG_AMOUNT         = "wrong_amount"
    const val VOUCHER_FAKE         = "voucher_fake"
    const val NO_RESPONSE          = "no_response"
    const val OTHER                = "other"

    /** Texto legible en español para mostrar en la UI */
    fun label(reason: String) = when (reason) {
        PAYMENT_NOT_RECEIVED -> "Pago no recibido"
        WRONG_AMOUNT         -> "Monto incorrecto"
        VOUCHER_FAKE         -> "Comprobante falso"
        NO_RESPONSE          -> "Sin respuesta"
        OTHER                -> "Otro motivo"
        else                 -> reason
    }

    val all = listOf(
        PAYMENT_NOT_RECEIVED,
        WRONG_AMOUNT,
        VOUCHER_FAKE,
        NO_RESPONSE,
        OTHER
    )
}

File: p2p_backend/app/models/dispute.py
"""Modelo Dispute — tabla disputes"""
from app.core.database import db, BaseModel


class Dispute(BaseModel):
    """
    Representa una disputa abierta sobre una transacción.

    Estados posibles:
        open       → recién abierta, esperando revisión admin
        under_review → admin la tomó, está revisando
        resolved   → resuelta (liberado al comprador o revertido al vendedor)
        closed     → cerrada sin acción (ej. partes llegaron a acuerdo)
    """
    __tablename__ = 'disputes'

    transaction_id  = db.Column(db.String(36), db.ForeignKey('transactions.id'),
                                nullable=False, index=True)
    initiator_id    = db.Column(db.String(36), db.ForeignKey('users.id'),
                                nullable=False)
    reason          = db.Column(db.String(255), nullable=False)
    description     = db.Column(db.Text)
    status          = db.Column(db.String(20), default='open', index=True)

    # Campos de resolución — se rellenan cuando admin resuelve
    resolved_by     = db.Column(db.String(36), db.ForeignKey('users.id'), nullable=True)
    resolution      = db.Column(db.String(20), nullable=True)   # 'favour_buyer' | 'favour_vendor'
    resolution_note = db.Column(db.Text, nullable=True)
    resolved_at     = db.Column(db.DateTime, nullable=True)

    # Relationships (lazy para evitar N+1 en listados)
    transaction = db.relationship('Transaction', foreign_keys=[transaction_id],
                                  backref=db.backref('disputes', lazy='dynamic'), lazy='joined')
    initiator   = db.relationship('User', foreign_keys=[initiator_id], lazy='joined')
    resolver    = db.relationship('User', foreign_keys=[resolved_by], lazy='select')

    # ── Razones de disputa predefinidas ──────────────────────────────────────
    REASON_PAYMENT_NOT_RECEIVED  = 'payment_not_received'
    REASON_WRONG_AMOUNT          = 'wrong_amount'
    REASON_VOUCHER_FAKE          = 'voucher_fake'
    REASON_NO_RESPONSE           = 'no_response'
    REASON_OTHER                 = 'other'

    VALID_REASONS = {
        REASON_PAYMENT_NOT_RECEIVED,
        REASON_WRONG_AMOUNT,
        REASON_VOUCHER_FAKE,
        REASON_NO_RESPONSE,
        REASON_OTHER,
    }

    def to_dict(self, include_transaction=False):
        data = {
            'id':              self.id,
            'transaction_id':  self.transaction_id,
            'initiator_id':    self.initiator_id,
            'initiator_name':  self.initiator.full_name if self.initiator else None,
            'reason':          self.reason,
            'description':     self.description,
            'status':          self.status,
            'resolved_by':     self.resolved_by,
            'resolution':      self.resolution,
            'resolution_note': self.resolution_note,
            'resolved_at':     self.resolved_at.isoformat() if self.resolved_at else None,
            'created_at':      self.created_at.isoformat(),
            'updated_at':      self.updated_at.isoformat(),
        }
        if include_transaction and self.transaction:
            t = self.transaction
            data['transaction'] = {
                'id':           t.id,
                'amount_from':  t.amount_from,
                'amount_to':    t.amount_to,
                'exchange_rate': t.exchange_rate,
                'status':       t.status,
                'buyer_id':     t.buyer_id,
                'vendor_id':    t.vendor_id,
            }
        return data
